src/libs/kata-sys-util/src/hooks.rs

```diff
@@ -0,0 +1,519 @@
+// Copyright (c) 2019-2021 Alibaba Cloud
+// Copyright (c) 2019-2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::collections::HashMap;
+use std::ffi::OsString;
+use std::hash::{Hash, Hasher};
+use std::io::{self, Read, Result};
+use std::path::Path;
+use std::time::Duration;
+
+use subprocess::{ExitStatus, Popen, PopenConfig, PopenError, Redirection};
+
+use crate::{eother, sl};
+
+const DEFAULT_HOOK_TIMEOUT_SEC: i32 = 10;
+
+#[derive(Debug, Default, Clone)]
+struct HookKey(oci::Hook);
+
+impl From<&oci::Hook> for HookKey {
+    fn from(hook: &oci::Hook) -> Self {
+        HookKey(hook.clone())
+    }
+}
+
+impl PartialEq for HookKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl Eq for HookKey {}
+
+impl Hash for HookKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.path.hash(state);
+        self.0.args.hash(state);
+        self.0.env.hash(state);
+        self.0.timeout.hash(state);
+    }
+}
+
+/// Execution state of OCI hooks.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum HookState {
+    /// Hook is pending for executing/retry.
+    Pending,
+    /// Hook has been successfully executed.
+    Done,
+    /// Hook has been marked as ignore.
+    Ignored,
+}
+
+/// Structure to maintain state for hooks.
+#[derive(Default)]
+pub struct HookStates {
+    states: HashMap<HookKey, HookState>,
+}
+
+impl HookStates {
+    /// Create a new instance of [`HookStates`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Get execution state of a hook.
+    pub fn get(&self, hook: &oci::Hook) -> HookState {
+        self.states
+            .get(&hook.into())
+            .copied()
+            .unwrap_or(HookState::Pending)
+    }
+
+    /// Update execution state of a hook.
+    pub fn update(&mut self, hook: &oci::Hook, state: HookState) {
+        self.states.insert(hook.into(), state);
+    }
+
+    /// Remove an execution state of a hook.
+    pub fn remove(&mut self, hook: &oci::Hook) {
+        self.states.remove(&hook.into());
+    }
+
+    /// Check whether some hooks are still pending and should retry execution.
+    pub fn should_retry(&self) -> bool {
+        for state in self.states.values() {
+            if *state == HookState::Pending {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Execute an OCI hook.
+    ///
+    /// If `state` is valid, it will be sent to subprocess' STDIN.
+    ///
+    /// The OCI Runtime specification states:
```

Comment:
Which version? Maybe you could just link to the exact version using an appropriate link?

Reply:
At least runtime 1.0.0 spec has this definition:
https://github.com/opencontainers/runtime-spec/releases/download/v1.0.0/oci-runtime-spec-v1.0.0.pdf

We add it to the comment.